### PR TITLE
Added support for the CRUD operations on the Item rest endpoint

### PR DESCRIPTION
**Pull request by KevinVdV**
In this PR we have created support for CRUD operations on the item endpoint for the REST api. This includes:

* A PUT method in the RestResourceController to be used for the repository.put methods which will take care of the updating. This is build with the createAndReturn example in mind and functions virtually in the same way.
* A small improvement on the DELETE method so that only admins can trigger an item delete
* The creation of the createAndReturn(create), put(update) methods in the ItemRestRepository
* We've added a class DSpaceObjectUtils that takes care of the replacing of the MetadataEntries so that we have one centralised place to manage this
* Additional tests has been added to the ItemRestRepositoryIT to keep the functionality properly tested at all times

Jira link: https://jira.duraspace.org/browse/DS-4108
RestContract PR link: DSpace/Rest7Contract#37

### CODE REVIEW

**Comment by KevinVdV** [comment]
This PR can be considered work in progress as it will have the same issues as the comm/coll crud: #2277

**Comment by KevinVdV** [comment]
The alterations requested in the contract have been applied to the codebase, this PR is ready for review.

**Review by tdonohue — COMMENTED** [review]
I did a quick code review today.  Some inline comments below, mostly regarding exceptions

File: dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java [hunk 1]
```diff
@@ -407,4 +408,44 @@ protected void patch(Context context, HttpServletRequest request, String apiCate
...
+     * @param jsonNode    the part of the request body representing the updated rest object
+     * @return the updated REST object
+     */
+    public T put(HttpServletRequest request, String apiCategory, String model, ID uuid, JsonNode jsonNode) {
+        Context context = obtainContext();
+        try {
+            thisRepository.put(context, request, apiCategory, model, uuid, jsonNode);
+            context.commit();
+        } catch (SQLException | AuthorizeException e) {
+            throw new RuntimeException(e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (mwoodiupui) [reply dspace-spring-rest/src/main/java/org/dspace/app/rest/repository/DSpaceRestRepository.java]: It is helpful for each layer of exception to explain what that code was trying (and failed) to do. For example, 'throw new RuntimeException("Could not update object " + uuid, e);'  The stack trace will show what caused that failure.

**Comment by KevinVdV** [comment]
Hi Tim,

Made the alterations you requested, however when adding these changes I noticed that the "UnprocessableEntityException" class didn't have a constructor with an exception. I added this one in this PR, but after this PR is merged I would suggest that we alter the existing code to not swallow this exception.

**Review by tdonohue — APPROVED** [review]
👍 Gave this another review, and it now looks good to me.  Thanks for the updates, @KevinVdV

**Review by terrywbrady — COMMENTED** [review]
@KevinVdV , The code changes look good.  Ping me after the conflicts are resolved and I will run a test.

**Comment by KevinVdV** [comment]
@terrywbrady This PR has been rebased on the master, so ready for testing !

**Comment by Raf-atmire** [comment]
We have removed the getName() check from the DspaceObjectRestEqualityUtils class because it didn't make sense to always have to provide the old title of the object with every request.
This has been fixed in the latest commit

**Review by terrywbrady — APPROVED** [review]
Code looks good.  I used Postman to verify the PUT operation and verified changes.

**Comment by tdonohue** [comment]
This is at +2. Merging.

**Pull request merged**